ykcompile/src/lib.rs

```diff
@@ -0,0 +1,175 @@
+#![feature(proc_macro_hygiene)]
+#![feature(test)]
+
+#[macro_use]
+extern crate dynasm;
+extern crate dynasmrt;
+extern crate test;
+
+use std::collections::HashMap;
+use std::mem;
+
+use yktrace::tir::{TirTrace, TirOp, Statement, Rvalue, Operand, Constant, ConstantInt,
+UnsignedInt};
+
+use dynasmrt::DynasmApi;
+
+/// A compiled SIRTrace.
+pub struct CompiledTrace {
+    /// A compiled trace.
+    mc: dynasmrt::ExecutableBuffer
+}
+
+impl CompiledTrace {
+    pub fn execute(&self) -> u64 {
+        // For now a compiled trace always returns whatever has been left in register RAX. We also
+        // assume for now that this will be a `u64`.
+        let func: fn() -> u64 = unsafe {
+            mem::transmute(self.mc.ptr(dynasmrt::AssemblyOffset(0)))
+        };
+        func()
+    }
+}
+
+/// The `TraceCompiler` takes a `SIRTrace` and compiles it to machine code. Returns a `CompiledTrace`.
+pub struct TraceCompiler {
+    asm: dynasmrt::x64::Assembler,
+    available_regs: Vec<u8>,
+    assigned_regs: HashMap<u32, u8>
+}
+
+impl TraceCompiler {
+
+    fn local_to_reg(&mut self, l: u32) -> u8 {
+        // This is a really dumb register allocator, which runs out of available registers after 7
+        // locals. We can do better than this by using StorageLive/StorageDead from the MIR to free
+        // up registers again, and allocate additional locals on the stack. Though, ultimately we
+        // probably want to implement a proper register allocator, e.g. graph colouring or linear
+        // scan.
+        match self.assigned_regs.get(&l) {
```

Comment:
I wonder if we can make this more concise using `Entry`?

https://doc.rust-lang.org/std/collections/hash_map/enum.Entry.html

Reply:
Even better, we can just use a `Vec` where the index is the local variable number. That removes the hashmap altogether.

But we can treat that as an optimisation later tbh.

Reply:
I'll try.

Reply:
> Even better, we can just use a `Vec` where the index is the local variable number. That removes the hashmap altogether.
> 
> But we can treat that as an optimisation later tbh.

Not really, since our locals are all over the place i.e. not consecutive. But since all of this is going to get replaced by a proper register allocator some day, there's no point spending time optimising things now.

Reply:
ok